src/Engine/ProtoCore/DSASM/Executive.cs

```diff
@@ -1513,8 +1513,15 @@ private void XLangUpdateDependencyGraph(int currentLangBlock)
                         // happens. 
                         if (graphNode.isLanguageBlock && currentLangBlock != Constants.kInvalidIndex)
                         {
-                            if (graphNode.languageBlockId == currentLangBlock
-                                || exe.CompleteCodeBlocks[currentLangBlock].IsMyAncestorBlock(graphNode.languageBlockId))
+                            if (graphNode.languageBlockId == currentLangBlock)
+                            {
+                                continue;
+                            }
+
+                            var cb = CoreUtils.GetCodeBlock(exe.CompleteCodeBlocks, currentLangBlock);
```

Comment:
Doesn't this still have the potential to be slow if the input size is large? Programs can be of arbitrarily large size can't they?

What about a version of `Coreutils.GetCodeBlock` that takes an initial index to check?

I guess eventually we'd want to stop using arrays and start using a sorted or balanced data structure with fast lookup?

Reply:
The whole issue is that codeBlockIds are not equivalent to the indexes in `CompleteCodeBlocks`.
A version of `Coreutils.GetCodeBlock` that takes an initial index to check could return wrong data (even though we could guard against invalid indexes)

As I mentioned in the description, I want to refactor these data structs and replace with something more suitable to the access patterns (ex. simple Dictionionary<codeBlockId, data> )

For the first question, yes there is potential to degrade if we have graphs with a large number of codeblocks (hundreds of thousands).
There is an opportunity to optimize the existing changes I made (like moving `CoreUtils.GetCodeBlock` outside the for in for loops). I would still leave it as it is even if it gets a bit slower. 
Otherwise we would have to change a lot of code to refactor in the current PR.

Reply:
thanks for the explanation - for a bit further clarification - when you say `have graphs with a large number of codeblocks (hundreds of thousands).` you are referring to codeblocks as `codeblocks` in the VM/AST right, not codeblock nodes?

IE - a single node might generate many codeblocks?

Reply:
IE - a single node might generate many codeblocks? 
Yes

Reply:
I am hoping to do the refactor as soon as possible. Do you think that because of possibility of perf degradation due to this PR would we should include the refactor here as well ?

Reply:
@mjkkirschner yes, `codeblock`s here refer to a code scope. For example the body of a for, while, if, else statements are all codeblocks by themselves. Each one of these codeblocks has a symbol table, procedure table, instruction stream, etc. 
For example, the following DS code snippet:
```
def foo(a, b)
{
	return [Imperative]
	{
		list = a..b;
		x = [];
		count = 0;
		for(i in list)
		{
			x[count] = i;
			count = count+1;
		}
		return x;
	}
};

x = true;
l1 = foo(1, 10);
l2 = foo(10, 20);

[Imperative]
{
	if(x==true)
	{
		return l1;
	}
	else
	{
		return l2;
	}
};
```
generates the following instruction set. Each instruction is prefixed with a number like [i.1.0], which is of the format ['associative: a/imperative: i', `codeblock index`, `instruction index`]:
```
def foo:var[]..[](a:var, b:var)
{
        ==============Start Node==============
        [i.1.0]push a
        [i.1.1]push b
        [i.1.2]push null
        [i.1.3]push 0
        [i.1.4]push False
        [i.1.5]push False
        [i.1.6]push 0[depth]
        [i.1.7]callr %generate_range
        [i.1.8]push _rx
        [i.1.9]setexpuid 19[exprId]
        [i.1.10]pop list
        [i.1.11]setexpuid 20[exprId]
        [i.1.12]newarr 0
        [i.1.13]pop x
        [i.1.14]setexpuid 21[exprId]
        [i.1.15]push 0
        [i.1.16]pop count
        [i.1.17]setexpuid 22[exprId]
        [i.2.18]pushb 2
        [i.2.19]push null
        [i.2.20]pop i
        [i.2.21]setexpuid 23[exprId]
        [i.2.22]push list
        [i.2.23]pop %forloop_expr_0
        [i.2.24]setexpuid 24[exprId]
        [i.2.25]push 0
        [i.2.26]pop %forloop_count_0
        [i.2.27]setexpuid 25[exprId]
        [i.2.28]pushvarsize %forloop_expr_0
        [i.2.29]pop %forloop_key_0
        [i.2.30]push %forloop_key_0
        [i.2.31]push null
        [i.2.32]nq
        [i.2.33]cjmp  L(-1)
        [i.2.34]setexpuid 26[exprId]
        [i.3.35]pushb 3
        [i.3.36]push %forloop_expr_0
        [i.3.37]push %forloop_count_0
        [i.3.38]push 0[depth]
        [i.3.39]callr ValueAtIndexInForLoop
        [i.3.40]push _rx
        [i.3.41]pop i
        [i.3.42]setexpuid 27[exprId]
        [i.3.43]push %forloop_key_0
        [i.3.44]push 1
        [i.3.45]add
        [i.3.46]pop %forloop_key_0
        [i.3.47]setexpuid 28[exprId]
        [i.3.48]push %forloop_count_0
        [i.3.49]push 1
        [i.3.50]add
        [i.3.51]pop %forloop_count_0
        [i.3.52]setexpuid 29[exprId]
        [i.3.53]push i
        [i.3.54]push count
        [i.3.55]push 1[dim]
        [i.3.56]setelement x
        [i.3.57]setexpuid 30[exprId]
        [i.3.58]push count
        [i.3.59]push 1
        [i.3.60]add
        [i.3.61]pop count
        [i.3.62]setexpuid 31[exprId]
        [i.3.63]retcn
        [i.2.64]jmp  L1(30)
        [i.2.65]retcn
        [i.1.66]push x
        [i.1.67]retb
        [i.1.68]setexpuid 32[exprId]
        [a.0.2]bounce 1, 0
        [a.0.3]push _rx
        [a.0.4]return
        ==============End Node==============
}

==============Start Node==============
[a.0.219]push True
[a.0.220]pop x
[a.0.221]dep 1[ExprUID] False[SSA]
[a.0.222]jdep
==============End Node==============
==============Start Node==============
[a.0.223]pushlevel @0
[a.0.224]popguides 0
[a.0.225]push 1
[a.0.226]pushlevel @0
[a.0.227]popguides 0
[a.0.228]push 10
[a.0.229]push 0[depth]
[a.0.230]callr foo
[a.0.231]push _rx
[a.0.232]pop %t_0_bfef6a96488243a381505b36007a5c28
[a.0.233]dep 2[ExprUID] True[SSA]
[a.0.234]jdep
==============End Node==============
==============Start Node==============
[a.0.235]push %t_0_bfef6a96488243a381505b36007a5c28
[a.0.236]pop l1
[a.0.237]dep 2[ExprUID] False[SSA]
[a.0.238]jdep
==============End Node==============
==============Start Node==============
[a.0.239]pushlevel @0
[a.0.240]popguides 0
[a.0.241]push 10
[a.0.242]pushlevel @0
[a.0.243]popguides 0
[a.0.244]push 20
[a.0.245]push 0[depth]
[a.0.246]callr foo
[a.0.247]push _rx
[a.0.248]pop %t_1_bfef6a96488243a381505b36007a5c28
[a.0.249]dep 3[ExprUID] True[SSA]
[a.0.250]jdep
==============End Node==============
==============Start Node==============
[a.0.251]push %t_1_bfef6a96488243a381505b36007a5c28
[a.0.252]pop l2
[a.0.253]dep 3[ExprUID] False[SSA]
[a.0.254]jdep
==============End Node==============
==============Start Node==============
[i.4.0]push x
[i.4.1]push True
[i.4.2]eq
[i.4.3]cjmp  L(-1)
[i.4.4]setexpuid 33[exprId]
[i.5.5]pushb 5
[i.5.6]push l1
[i.5.7]retb
[i.5.8]setexpuid 34[exprId]
[i.5.9]retcn
[i.4.10]jmp  L1(-1)
[i.6.11]pushb 6
[i.6.12]push l2
[i.6.13]retb
[i.6.14]setexpuid 35[exprId]
[i.6.15]retcn
[i.4.16]push null
[i.4.17]retb
[a.0.255]bounce 4, 0
[a.0.256]push _rx
[a.0.257]pop %tempLangBlock9
[a.0.258]dep -1[ExprUID] False[SSA]
[a.0.259]jdep
==============End Node==============
[a.0.260]push null
[a.0.261]retb
```